src/inhomcorr/methods/unet.py

```diff
@@ -0,0 +1,228 @@
+"""Implementation of a UNet."""
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+from inhomcorr.interfaces.bias_estimator_interface import HyperParameters
+
+
+class ConvBlock(nn.Module):
+    """A block of convolutional layers (1D, 2D or 3D)."""
+
+    def __init__(
+            self,
+            dim,
+            n_ch_in,
+            n_ch_out,
+            n_convs,
+            kernel_size=3,
+            bias=False):
+        super(ConvBlock, self).__init__()
+
+        if dim == 1:
+            conv_op = nn.Conv1d
+        if dim == 2:
+            conv_op = nn.Conv2d
+        elif dim == 3:
+            conv_op = nn.Conv3d
+
+        padding = int(np.floor(kernel_size / 2))
+
+        conv_block_list = []
+        conv_block_list.extend([conv_op(n_ch_in,
+                                        n_ch_out,
+                                        kernel_size,
+                                        padding=padding,
+                                        bias=bias),
+                                nn.LeakyReLU()])
+
+        for i in range(n_convs - 1):
+            conv_block_list.extend([conv_op(
+                n_ch_out, n_ch_out, kernel_size, padding=padding, bias=bias),
+                nn.LeakyReLU()])
+
+        self.conv_block = nn.Sequential(*conv_block_list)
+
+    def forward(self, x):
+        """Forward pass of the convolutional block.
+
+        Parameters
+        ----------
+        x
+            _description_
+
+        Returns
+        -------
+            _description_
+        """
+        return self.conv_block(x)
+
+
+class Encoder(nn.Module):
+    """Encoder of the UNet."""
+
+    def __init__(
+            self,
+            dim,
+            n_ch_in,
+            n_enc_stages,
+            n_convs_per_stage,
+            n_filters,
+            kernel_size=3,
+            bias=False):
+        super(Encoder, self).__init__()
+
+        n_ch_list = [n_ch_in]
+        for ne in range(n_enc_stages):
+            n_ch_list.append(int(n_filters) * 2**ne)
+
+        self.enc_blocks = nn.ModuleList([ConvBlock(dim,
+                                                   n_ch_list[i],
+                                                   n_ch_list[i + 1],
+                                                   n_convs_per_stage,
+                                                   kernel_size=kernel_size)
+                                         for i in range(len(n_ch_list) - 1)])
+
+        if dim == 1:
```

Comment:
do we want to fix our self to 2d for sake of simplicity? and use a different class for 3d etc?